tokio-trace/tokio-trace-core/src/subscriber.rs

```diff
@@ -0,0 +1,353 @@
+    /// identity function, passing through the identifier. However, it can be
+    /// used in conjunction with [`drop_span`] to track the number of handles
+    /// capable of `enter`ing a span. When all the handles have been dropped
+    /// (i.e., `drop_span` has been called one more time than `clone_span` for a
+    /// given ID), the subscriber may assume that the span will not be entered
+    /// again. It is then free to deallocate storage for data associated with
+    /// that span, write data from that span to IO, and so on.
+    ///
+    /// For more unsafe situations, however, if `id` is itself a pointer of some
+    /// kind this can be used as a hook to "clone" the pointer, depending on
+    /// what that means for the specified pointer.
+    ///
+    /// [`Id`]: ::span::Id,
+    /// [`drop_span`]: ::subscriber::Subscriber::drop_span
+    fn clone_span(&self, id: &Span) -> Span {
+        id.clone()
+    }
+
+    /// Notifies the subscriber that a [`Span`] handle with the given [`Id`] has
+    /// been dropped.
+    ///
+    /// This function is guaranteed to only be called with span IDs that were
+    /// returned by this subscriber's `new_span` function.
+    ///
+    /// It's guaranteed that if this function has been called once more than the
+    /// number of times `clone_span` was called with the same `id`, then no more
+    /// `Span`s using that `id` exist. This means that it can be used in
+    /// conjunction with [`clone_span`] to track the number of handles
+    /// capable of `enter`ing a span. When all the handles have been dropped
+    /// (i.e., `drop_span` has been called one more time than `clone_span` for a
+    /// given ID), the subscriber may assume that the span will not be entered
+    /// again. It is then free to deallocate storage for data associated with
+    /// that span, write data from that span to IO, and so on.
+    ///
+    /// **Note**: since this function is called when spans are dropped,
+    /// implementations should ensure that they are unwind-safe. Panicking from
+    /// inside of a `drop_span` function may cause a double panic, if the span
+    /// was dropped due to a thread unwinding.
+    ///
+    /// [`Id`]: ::span::Id,
+    /// [`drop_span`]: ::subscriber::Subscriber::drop_span
+    fn drop_span(&self, id: Span) {
+        let _ = id;
+    }
+}
+
+/// Indicates a `Subscriber`'s interest in a particular callsite.
+#[derive(Clone, Debug)]
+pub struct Interest(InterestKind);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum InterestKind {
+    Never = 0,
+    Sometimes = 1,
+    Always = 2,
+}
+
+impl Interest {
+    /// Indicates that the subscriber is never interested in being notified
+    /// about a callsite.
+    ///
+    /// If all active subscribers are `NEVER` interested in a callsite, it will
+    /// be completely disabled unless a new subscriber becomes active.
+    pub const NEVER: Interest = Interest(InterestKind::Never);
```

Comment:
Thanks for bringing up the potential forward-compatibility hazard of using consts, I hadn't thought of that. I think these weren't functions because I wanted to avoid having to potentially push a stack frame to construct a new interest, but they can probably be changed.